### PR TITLE
Resource Loader - Better Activation Type, NIO-resource pack creation methods, and fixes

### LINKED ISSUE #242: API for constructing a resource pack from a path
There's a lot of knowledge and work tied up in QSL Resource Loader; it's a very solid library that powers loading of resources in mods. As a result of frustrations I and many others have had with existing "resource loader" mods, the other day I ended up writing and releasing [QDResLoader](https://modrinth.com/mod/qdresloader) — a positively tiny mod that leverages QSL Resource Loader to provide a loader mod that has minimal complexity of its own. Building on the shoulders of giants, so to speak.

However, doing this requires using private API — it directly constructs ModNioResourcePack instances. Thankfully, an event for registering raw ResourcePack instances directly got added in #191, ~~and QDResLoader leverages that~~ (EDIT: Not anymore, I had to change to a mixin because `Default` is too early and `Top` is too late. See #243). But I have no choice but to reinvent the wheel on top of MutableResourcePack, or use private API. As the whole point of this was to reuse QSL's battletested systems, I ended up using the private API.

I think it would be useful if there were an API method to construct a ResourcePack from a Path, completely generically, internally constructing a ModNioResourcePack (or a rename and generalization thereof). I've also wanted this in the past for providing convenient item/recipe generation and definition in other mods of mine, but that's covered fairly well by the MutableResourcePack API now — nonetheless, I did wind up implementing that by directly calling into ModNioResourcePack back then as well.

While I could just copy the entirety of ModNioResourcePack into my mod (it is not a very large class, even including its various helper and utility classes), this is kind of silly, and would not benefit from further work on QSL Resource Loader without manually pulling in changes. `ModResourcePackUtil` is itself internal API as well.

### PR DESCRIPTION
**Pull request by LambdAurora**
This PR aims to resolve #242 and partially #243.

The event discussed in #243 might need more thinking and discussion before making into a PR.

### CODE REVIEW

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/ResourcePackProfileMixin.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+	@Inject(
+			method = "of",
+			at = @At(
+					value = "INVOKE",
+					target = "Lnet/minecraft/resource/pack/ResourcePackProfile$Factory;create(Ljava/lang/String;Lnet/minecraft/text/Text;ZLjava/util/function/Supplier;Lnet/minecraft/resource/pack/metadata/PackResourceMetadata;Lnet/minecraft/resource/pack/ResourcePackProfile$InsertionPosition;Lnet/minecraft/resource/pack/ResourcePackSource;)Lnet/minecraft/resource/pack/ResourcePackProfile;"
+			),
+			locals = LocalCapture.CAPTURE_FAILHARD
+	)
+	private static void onProfileCreation(String name, boolean alwaysEnabled, Supplier<ResourcePack> packFactory,
```
Comment (EnnuiL) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/ResourcePackProfileMixin.java]: Hm, with PR #241 active again, why not begin to apply it now?
```suggestion
	private static void quilt$onProfileCreation(String name, boolean alwaysEnabled, Supplier<ResourcePack> packFactory,
```

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/ResourcePackProfileMixin.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+			ResourcePackProfile.Factory profileFactory, ResourcePackProfile.InsertionPosition insertionPosition, ResourcePackSource packSource,
+			CallbackInfoReturnable<ResourcePackProfile> cir, ResourcePack pack) {
+		quilt$KNOWN_ACTIVATION_TYPE.set(alwaysEnabled ? ResourcePackActivationType.ALWAYS_ENABLED : pack.getActivationType());
+	}
+
+	@Inject(
+			method = "<init>(Ljava/lang/String;ZLjava/util/function/Supplier;Lnet/minecraft/text/Text;Lnet/minecraft/text/Text;Lnet/minecraft/resource/pack/ResourcePackCompatibility;Lnet/minecraft/resource/pack/ResourcePackProfile$InsertionPosition;ZLnet/minecraft/resource/pack/ResourcePackSource;)V",
+			at = @At("RETURN")
+	)
+	private void onInit(String name, boolean alwaysEnabled, Supplier<ResourcePack> packFactory, Text displayName, Text description,
```
Comment (EnnuiL) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/ResourcePackProfileMixin.java]:
```suggestion
	private void quilt$onInit(String name, boolean alwaysEnabled, Supplier<ResourcePack> packFactory, Text displayName, Text description,
```

**Review by EnnuiL — APPROVED** [review]
Looks good! Only few tiny nitpicks that don't have to be applied now:

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/ResourcePackProfileMixin.java (hunk 1 above)
Comment (LambdAurora) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/ResourcePackProfileMixin.java]: I honestly feel like this is a band-aid solution to the actual problem.
Mixin should be fixed directly.

Comment (EnnuiL) [reply library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/ResourcePackProfileMixin.java]: hm, fair enough

**Review by TheGlitch76 — APPROVED** [review]

**Comment by exaskye** [comment]
QDResLoader now has a [prototype branch](https://git.sleeping.town/unascribed/QDResLoader/src/branch/qsl-pr-244) making use of this PR — it seems to work, and fixes a variety of bugs and shortcomings in the current implementation. :+1:

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoader.java [hunk 3]
```diff
@@ -98,6 +102,83 @@ public interface ResourceLoader {
...
+			ResourcePackActivationType activationType) {
+		return this.newFileSystemResourcePack(id, rootPath, activationType, ResourceLoaderImpl.getBuiltinPackDisplayNameFromId(id));
+	}
+
+	/**
+	 * Creates a new resource pack based on a {@link Path} as its root.
+	 * <p>
+	 * If the file system of the given {@link Path} is not the operating system's file system then the pack resources will be cached.
+	 *
+	 * @param id             the identifier of the resource pack; its namespace must be the same as the mod id
```
Comment (Leo40Git) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoader.java]:
```suggestion
	 * @param id             the identifier of the resource pack; its namespace must be the same as the mod ID
```

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoader.java [hunk 4]
```diff
@@ -98,6 +102,83 @@ public interface ResourceLoader {
...
+	 * @return a new resource pack instance
+	 * @see #newFileSystemResourcePack(Identifier, Path, ResourcePackActivationType)
+	 * @see #newFileSystemResourcePack(Identifier, ModContainer, Path, ResourcePackActivationType)
+	 * @see #newFileSystemResourcePack(Identifier, ModContainer, Path, ResourcePackActivationType, Text)
+	 */
+	default @NotNull ResourcePack newFileSystemResourcePack(@NotNull Identifier id, @NotNull Path rootPath,
+			ResourcePackActivationType activationType, @NotNull Text displayName) {
+		var container = QuiltLoader.getModContainer(id.getNamespace())
+				.orElseThrow(() ->
+						new IllegalArgumentException("No mod with mod id " + id.getNamespace() + " could be found"));
```
Comment (Leo40Git) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoader.java]:
```suggestion
						new IllegalArgumentException("No mod with ID '" + id.getNamespace() + "' could be found"));
```

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoader.java [hunk 5]
```diff
@@ -98,6 +102,83 @@ public interface ResourceLoader {
 	@Contract(pure = true)
 	@NotNull Event<ResourcePackRegistrationContext.Callback> getRegisterTopResourcePackEvent();
 
+	/**
+	 * Creates a new resource pack based on a {@link Path} as its root.
+	 * <p>
+	 * If the file system of the given {@link Path} is not the operating system's file system then the pack resources will be cached.
+	 *
+	 * @param id             the identifier of the resource pack; its namespace must be the same as the mod id
```
Comment (Leo40Git) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoader.java]:
```suggestion
	 * @param id             the identifier of the resource pack; its namespace must be the same as the mod ID
```

**Review by Leo40Git — APPROVED** [review]
LGTM, except for these 3 tiny things.

**Pull request merged**